File: packages/pluto/src/tester.ts
import {
  FnResource,
  IResource,
  IResourceClientApi,
  IResourceInfraApi,
  PlatformType,
  simulator,
  utils,
} from "@plutolang/base";

export interface TestCase {
  description: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  testHandler: any; // Not using pulumi.Output to avoid depending on pulumi.
}

export interface TestHandler extends FnResource {
  (): Promise<void>;
}

/**
 * The options for instantiating an infrastructure implementation class or a client implementation
 * class.
 */
export interface TesterOptions {}

/**
 * Don't export these methods to developers.
 * These methods are only used internally by the cli.
 */
export interface ITesterClientApi extends IResourceClientApi {
  listTests(): Promise<TestCase[]>;
  runTest(testCase: TestCase): Promise<void>;
}

export interface ITesterInfraApi extends IResourceInfraApi {
  it(description: string, fn: TestHandler): void;
}

export interface ITesterCapturedProps extends IResourceInfraApi {}

/**
 * Construct a type that includes all the necessary methods required to be implemented within the
 * client implementation class of a resource type.
 */
export type ITesterClient = ITesterClientApi & ITesterCapturedProps;

/**
 * Construct a type that includes all the necessary methods required to be implemented within the
 * infrastructure implementation class of a resource type.
 */
export type ITesterInfra = ITesterInfraApi & ITesterCapturedProps;

export class Tester implements IResource {
  constructor(name: string, opts?: TesterOptions) {
    name;
    opts;
    throw new Error(
      "Cannot instantiate this class, instead of its subclass depending on the target runtime."
    );
  }

  public static buildClient(name: string, opts?: TesterOptions): ITesterClient {
    const platformType = utils.currentPlatformType();
    switch (platformType) {
      case PlatformType.Simulator: {
        if (!process.env.PLUTO_SIMULATOR_URL) throw new Error("PLUTO_SIMULATOR_URL doesn't exist");
        const resourceId = utils.genResourceId(Tester.fqn, name);
        return simulator.makeSimulatorClient(process.env.PLUTO_SIMULATOR_URL!, resourceId);
      }
      default:
        throw new Error(`not support this runtime '${platformType}'`);
    }
    opts;
  }

  public static fqn = "@plutolang/pluto.Tester";
}

export interface Tester extends IResource, ITesterClient, ITesterInfra {}

File: packages/pluto-py/pluto_client/tester.py
from dataclasses import dataclass
from typing import Any, Callable, Optional, List
from pluto_base.resource import (
    IResource,
    IResourceCapturedProps,
    IResourceClientApi,
    IResourceInfraApi,
)
from pluto_base.platform import PlatformType
from pluto_base import utils
from .utils import create_simulator_client


class TestCase:
    def __init__(self, description: str, test_handler: Any):
        self.description = description
        self.testHandler = test_handler


TestHandler = Callable[[], None]


@dataclass
class TesterOptions:
    pass


class ITesterClientApi(IResourceClientApi):
    def list_tests(self) -> List[TestCase]:
        raise NotImplementedError

    def run_test(self, test_case: TestCase) -> None:
        raise NotImplementedError


class ITesterInfraApi(IResourceInfraApi):
    def it(self, description: str, fn: TestHandler) -> None:
        raise NotImplementedError


class ITesterCapturedProps(IResourceCapturedProps):
    pass


class ITesterClient(ITesterClientApi, ITesterCapturedProps):
    pass


class ITesterInfra(ITesterInfraApi, ITesterCapturedProps):
    pass


class Tester(IResource, ITesterClient, ITesterInfra):
    fqn = "@plutolang/pluto.Tester"

    def __init__(self, name: str, opts: Optional[TesterOptions] = None):
        raise NotImplementedError(
            "Cannot instantiate this class, instead of its subclass depending on the target runtime."
        )

    @staticmethod
    def build_client(name: str, opts: Optional[TesterOptions] = None) -> ITesterClient:
        platform_type = utils.current_platform_type()
        if platform_type == PlatformType.Simulator:
            resource_id = utils.gen_resource_id(Tester.fqn, name)
            return create_simulator_client(resource_id)
        else:
            raise ValueError(f"not support this runtime '{platform_type}'")
